### PR TITLE
internal: Migrate ValidateProviderConfig RPC to fromproto6, fwserver, and toproto6

### PR DESCRIPTION
**Pull request by bflad**
Reference: https://github.com/hashicorp/terraform-plugin-framework/issues/215

One particular quirk that this implementation shows is that it is not entirely possible to avoid working with `tfsdk`/`fwserver` bits in `proto6server` as the current data types within the framework (`Config`, `Plan`, `State`) require the `Schema` so data can be appropriately converted from the protocol to the framework. To support this effort and prevent repetitively executing the provider defined `GetSchema` method, a `fwserver` method is introduced that caches the `Schema` and `Diagnostics` results on first use. This methodology can also be applied in the future to resolve the existing https://github.com/hashicorp/terraform-plugin-framework/issues/299.

### CODE REVIEW

File: internal/fwserver/server.go [hunk 1]
```diff
@@ -9,4 +14,41 @@ import (
...
+
+// ProviderSchema returns the Schema associated with the Provider. The Schema
+// and Diagnostics are cached on first use.
+func (s *Server) ProviderSchema(ctx context.Context) (*tfsdk.Schema, diag.Diagnostics) {
+	logging.FrameworkTrace(ctx, "Checking ProviderSchema lock")
+	s.providerSchemaMutex.Lock()
+
+	if s.providerSchema != nil {
+		s.providerSchemaMutex.Unlock()
+		return s.providerSchema, nil
```
Comment (detro) [inline internal/fwserver/server.go:41]: Should this return also `s.ProviderSchemaDiags`?

Unless the intention is to report potential diag-reported warnings/errors, only the first time?

File: internal/fwserver/server.go [hunk 2]
```diff
@@ -9,4 +14,41 @@ import (
...
+}
+
+// ProviderSchema returns the Schema associated with the Provider. The Schema
+// and Diagnostics are cached on first use.
+func (s *Server) ProviderSchema(ctx context.Context) (*tfsdk.Schema, diag.Diagnostics) {
+	logging.FrameworkTrace(ctx, "Checking ProviderSchema lock")
+	s.providerSchemaMutex.Lock()
+
+	if s.providerSchema != nil {
+		s.providerSchemaMutex.Unlock()
```
Comment (detro) [inline internal/fwserver/server.go:40]: To ensure future code doesn't break the needs of the mutex to be unlocked, I have seen a slightly different usage pattern (that I like). Something like:

```
s.ProviderSchemaMutex.Lock()
defer s.ProviderSchemaMutex.Unlock()
...
```

This way, whatever branching of the logic we take (the memoized one or not), the locking bit is handled safely and consistently.

**Review by detro — APPROVED** [review]
Just a suggestion around the way the mutex gets unlocked.

Mostly is moving things to the new structure, so LGTM.

File: internal/fwserver/server.go (hunk 2 above)
Comment (bflad) [reply internal/fwserver/server.go:40]: Thank you for this suggestion, it's a great one and one I'll implement!

File: internal/fwserver/server.go (hunk 1 above)
Comment (bflad) [reply internal/fwserver/server.go:41]: Yes, it should, thanks for noticing!

**Pull request merged**